virttest/utils_libvirt/libvirt_disk.py

```diff
@@ -119,3 +127,289 @@ def create_custom_metadata_disk(disk_path, disk_format,
+        process.run(link_cmd, shell=True, ignore_status=False)
+        if index == 0:
+            continue
+        else:
+            backing_key = syslink_folder_list[index-1]
+            backing_file_dict[key] = "../%s/%s.img" % (backing_key, backing_key)
+    # Create system link backing files
+    disk_format = origin_image_format
+    for key, value in list(backing_file_dict.items()):
+        backing_file_path = os.path.join(root_dir, key)
+        cmd = ("cd %s && qemu-img create -f %s -o backing_file=%s,backing_fmt=%s %s.img"
+               % (backing_file_path, created_image_format, value, disk_format, key))
+        process.run(cmd, shell=True, ignore_status=False)
+        disk_format = created_image_format
+    return os.path.join(backing_file_path, "%s.img" % syslink_folder_list[-1]), list(backing_file_dict.values())
+
+
+def do_blockcommit_repeatedly(vm, device_target, options, repeated_counts):
+    """
+    Do blockcommit repeatedly
+
+    :param vm: VM instance
+    :param device_target: device target
+    :param options: blockcommit options
+    :param repeated_counts: repeated counts for executing blockcommit
+    """
+    for count in range(0, repeated_counts):
+        virsh.blockcommit(vm.name, device_target,
+                          options, ignore_status=False, debug=True)
+
+
+def make_external_disk_snapshots(vm, device_target, postfix_n, snapshot_take):
+    """
+    Make external snapshots for disks only.
+
+    :param vm: VM instance
+    :param device_target: device target
+    :param postfix_n: postfix option
+    :param snapshot_take: snapshots taken.
+    :return: list containing absolute root path of snapshot files
+    """
+    first_disk_source = get_first_disk_source(vm)
+    root_dir = os.path.dirname(first_disk_source)
+    basename = os.path.basename(first_disk_source)
+    disk = device_target
+    external_snapshot_disks = []
+    # Make external snapshots for disks only
+    for count in range(1, snapshot_take + 1):
+        options = "%s_%s %s%s-desc " % (postfix_n, count,
+                                        postfix_n, count)
+        options += "--disk-only "
+        diskname = basename.split(".")[0]
+        snap_name = "%s.%s%s" % (diskname, postfix_n, count)
+        disk_external = os.path.join(root_dir, snap_name)
+        external_snapshot_disks.append(disk_external)
+        options += " %s,snapshot=external,file=%s" % (disk,
+                                                      disk_external)
+        virsh.snapshot_create_as(vm.name, options,
+                                 ignore_status=False,
+                                 debug=True)
+    return external_snapshot_disks
+
+
+def cleanup_snapshots(vm, snap_del_disks=None):
+    """
+    clean up snapshots
+
+    :param vm: VM instance
+    :param snap_del_disks: list containing snapshot files
+    """
+    snapshot_list_cmd = "virsh snapshot-list %s --tree" % vm.name
+    result_output = process.run(snapshot_list_cmd,
```

Comment:
Could we use virsh.snapshot_list instead?

Reply:
@kylazhang 
Primitive virsh.snapshot-list in virsh.py hardcode output parser, which can not deal with output correctly if --tree option is used
```
def snapshot_list(name, options=None, **dargs):
    """
    Get list of snapshots of domain.

    :param name: name of domain
    :param options: options of snapshot_list
    :param dargs: standardized virsh function API keywords
    :return: list of snapshot names
    """
    # CmdResult is handled here, force ignore_status
    dargs['ignore_status'] = True
    ret = []
    cmd = "snapshot-list %s" % name
    if options is not None:
        cmd += " %s" % options

    sc_output = command(cmd, **dargs)
    if sc_output.exit_status != 0:
        raise process.CmdError(
            cmd, sc_output, "Failed to get list of snapshots")

    data = re.findall("\S* *\d*-\d*-\d* \d*:\d*:\d* [+-]\d* \w*",
                      sc_output.stdout_text)
    for rec in data:
        if not rec:
            continue
        ret.append(re.match("\S*", rec).group())

    return ret

```

Reply:
@kylazhang
Primitive virsh.snapshot-list in virsh.py hardcode output parser, which can not deal with output correctly if --tree option is used

def snapshot_list(name, options=None, **dargs):
    """
    Get list of snapshots of domain.

    :param name: name of domain
    :param options: options of snapshot_list
    :param dargs: standardized virsh function API keywords
    :return: list of snapshot names
    """
    # CmdResult is handled here, force ignore_status
    dargs['ignore_status'] = True
    ret = []
    cmd = "snapshot-list %s" % name
    if options is not None:
        cmd += " %s" % options

    sc_output = command(cmd, **dargs)
    if sc_output.exit_status != 0:
        raise process.CmdError(
            cmd, sc_output, "Failed to get list of snapshots")

    data = re.findall("\S* *\d*-\d*-\d* \d*:\d*:\d* [+-]\d* \w*",
                      sc_output.stdout_text)
    for rec in data:
        if not rec:
            continue
        ret.append(re.match("\S*", rec).group())

    return ret

By the way, it is old codes existed in tp-libvirt already, but just move it to avocado-vt

Reply:
OK, get the reason.